diff -ur internal/race/norace.go /usr/local/Cellar/go/1.13.4/libexec/src/internal/race/norace.go
--- internal/race/norace.go	2020-01-10 15:17:12.000000000 -0600
+++ /usr/local/Cellar/go/1.13.4/libexec/src/internal/race/norace.go	2020-02-19 10:47:44.000000000 -0600
@@ -18,6 +18,12 @@
 func Release(addr unsafe.Pointer) {
 }
 
+func PostLock(addr unsafe.Pointer, write bool) {
+}
+
+func PreUnlock(addr unsafe.Pointer, write bool) {
+}
+
 func ReleaseMerge(addr unsafe.Pointer) {
 }
 
diff -ur internal/race/race.go /usr/local/Cellar/go/1.13.4/libexec/src/internal/race/race.go
--- internal/race/race.go	2020-01-10 15:17:12.000000000 -0600
+++ /usr/local/Cellar/go/1.13.4/libexec/src/internal/race/race.go	2020-02-18 21:21:30.000000000 -0600
@@ -21,6 +21,14 @@
 	runtime.RaceRelease(addr)
 }
 
+func PostLock(addr unsafe.Pointer, write bool) {
+	runtime.RacePostLock(addr, write)
+}
+
+func PreUnlock(addr unsafe.Pointer, write bool) {
+	runtime.RaceUnlock(addr, write)
+}
+
 func ReleaseMerge(addr unsafe.Pointer) {
 	runtime.RaceReleaseMerge(addr)
 }
